### ISSUE TITLE
Sampled Images are Noisy

### ISSUE THREAD

**Issue by ruchaa0112** [VSehwag/minimal-diffusion#1]
@VSehwag  ,
Thank you for your work. I really love the way the repo is structured - made it easy to understand and write custom data class. 
I am trying to train on a custom image data. I wrote a Dataloader for the custom dataset. On plotting the images from dataloader, they seem to appear correctly. I have also ensured that the pixel values are between [-1,1]. I am using single GPU for training. 
The training and sampling process seems to run , however the first set of images generated is complete noise. 
Here are the details about the training process - 
`
Training dataset loaded: Number of batches: 400, Number of images: 51200
Number of Parameters 32846977
Steps: 1/4000    loss (ema): 0.998       Time elapsed: 0.002 hr
Steps: 101/4000          loss (ema): 0.040       Time elapsed: 0.032 hr
Steps: 201/4000          loss (ema): 0.004       Time elapsed: 0.063 hr
Steps: 301/4000          loss (ema): 0.003       Time elapsed: 0.094 hr`

![UNet_poisson_2d-1000_steps-100-sampling_steps-class_condn_False](https://user-images.githubusercontent.com/124633423/222809152-865c1b26-c9bb-4c1c-8d38-f60789daa5d7.png)

Above is the output for sampling after 10 epochs. (Gray bar)

Can you please let me know if I am going wrong somewhere?

**Comment by VSehwag** [comment]
Unless it's a very easy dataset like mnist, the first epoch would generated very noisy images. For early 2-5 epochs, generated samples tend to be very blurry and just have low-frequency details. It take a lot of epochs of refine these samples to very high quality ones.

However, in this case your training loader is doing something wrong. Why? Because the MSE loss is 0.004 just after 200 steps. Even on mnist, mse loss would reduce to 0.02 after 10000 steps. So no way the loss could be this low, unless your training data is just a set of gray images (for which the denoising is straightforward and mse loss would be indeed low). So I urge you to save the images right before they are passed to the UNet and see if there is something wrong there.

**Comment by ruchaa0112** [comment]
Thank you so much for your prompt response ! 
I replicated the model with MNIST and seems to work fine. 
The dataset I am using is really very simple one for starters , governed by a mathematical equation. 

Can that be the reason for decrease of the MSE loss so rapidly?

**Comment by VSehwag** [comment]
Imo the issue is most likely with the dataloader. As suggested earlier, I recommend saving images right before they are passed to the UNet and see if there is something wrong there.

**Comment by ruchaa0112** [comment]
Thank you for your suggestion! I tried to plot the images before sending them to forward diffusion as the output of the dataloader, and they look correct to me. The loss seems to still rapidly decrease at the 2nd epoch itself, and the synthetic images generated are noise. 
Attached are the input images plotted just before the UNet process in train_one_epoch()
![75364a36-cd2a-4317-b4b4-384faaa5f66a](https://user-images.githubusercontent.com/124633423/223284719-d1e0ae72-cadf-415e-b37e-ddef2969e11e.jpg)
